MathlibTest/Isolate/Lightweight.lean

```diff
@@ -0,0 +1,207 @@
+example {x y z : ℝ} (_hy : 0 < y) (H : z < (x + 3) * y ^ 2 - 2) : x + 3 - 4 < y := by
+  isolate x + 3 at *
+  guard_hyp H : (z + 2) / y ^ 2 < x + 3
+  guard_target = x + 3 < y + 4
+  exact test_sorry
+
+-- isolate term specified by a pattern
+example {x y z : ℝ} (_hy : 0 < y) : (x + 3) * y ^ 2 - 2 = z := by
+  isolate _ + _
+  guard_target = x + 3 = (z + 2) / y ^ 2
+  exact test_sorry
+
+-- isolate term specified by a pattern
+example {x y : ℝ} : x ^ 2 + 3 = y := by
+  isolate _ ^ 2
+  guard_target = x ^ 2 = y - 3
+  exact test_sorry
+
+-- isolate a term whose elaboration must be delayed, so as to avoid choosing a default instance
+example {x y : ℝ} : x + 3 = y := by
+  isolate 3
+  guard_target = 3 = y - x
+  exact test_sorry
+
+-- isolate on the RHS of a symmetric relation
+example {x y z : ℝ} (_hy : 0 < y) : z = (x + 3) * y ^ 2 - 2 := by
+  isolate x + 3
+  guard_target = x + 3 = (z + 2) / y ^ 2
+  exact test_sorry
+
+-- isolation of an expression will proceed as far as possible, even if the expression cannot be
+-- fully isolated
+example {x : ℕ} {y : ℚ} : ↑(x + 2) + 3 = y := by
+  isolate x
+  guard_target = ↑(x + 2) = y - 3
+  exact test_sorry
+
+-- a term can be "isolated" even if it doesn't contain free variables
+example {x y : ℝ} : x + 3 = y := by
+  isolate (3:ℝ)
+  guard_target = 3 = y - x
+  exact test_sorry
+
+-- check context is correct
+example (a : ℚ) : True := by
+  have z : ℚ := test_sorry
+  have : z - 3 ≤ a := by
+    isolate z
+    exact test_sorry
+  exact trivial
+
+/-- error: x + 3 should appear in only one (not both) of x + 3 and (x + 3) * y ^ 2 - 2 -/
+#guard_msgs in
+example {x y z : ℝ} (_hy : 0 < y) : x + 3 < (x + 3) * y ^ 2 - 2 := by
+  isolate x + 3
+
+/-- error: x + 2 should appear in either z or (x + 3) * y ^ 2 - 2 -/
+#guard_msgs in
+example {x y z : ℝ} (_hy : 0 < y) : z < (x + 3) * y ^ 2 - 2 := by
+  isolate x + 2
+
+/-- error: No x + 2 terms found anywhere which could be isolated -/
+#guard_msgs in
+example {x y z : ℝ} (_hy : 0 < y) : z < (x + 3) * y ^ 2 - 2 := by
+  isolate x + 2 at *
+
+/-- error: x + 3 is already isolated in x + 3 = y -/
+#guard_msgs in
+example {x y : ℝ} : x + 3 = y := by
+  isolate x + 3
+
+/-- error: f should be a concrete function, for example it cannot be a variable -/
+#guard_msgs in
+example (x : ℝ) (f : ℝ → ℝ) : f x = 12 := by
+  isolate x
+
+/-! ### Tagging -/
+
+/-- info: [Mathlib.Tactic.Isolate.add_right_le_iff] -/
+#guard_msgs in
+#query_isolate_lemmas LE.le HAdd.hAdd 4 0
+
+/--
+error: @[isolate] attribute only applies to lemmas with a conclusion of the form f a₁ a₂ ... x ... aₖ ~ y ↔ x ~' G.
+ If-and-only-if structure not identified in this lemma's conclusion a * (b + c) = a * b + a * c
+-/
+#guard_msgs in
+attribute [isolate] mul_add
+
+axiom Prime : Nat → Prop
+
+/--
+error: @[isolate] attribute only applies to lemmas with a conclusion of the form f a₁ a₂ ... x ... aₖ ~ y ↔ x ~' G.
+ Here the conclusion has the form Prime a ↔ _, but Prime a could not be parsed as a relation
```

Comment:
I didn't really understand "a relation" here, maybe something like:

```suggestion
 Here the conclusion has the form Prime a ↔ _, but Prime a could not be parsed as a relation
 of the form `f a₁ a₂ ... x ... aₖ ~ y`?
```

Reply:
Addressed in 95317b6b274 - error message now clarifies that a relation must have the form `_ ~ _`